### PR TITLE
Switch to Trolley for Modrinth Payments

### PR DESCRIPTION
**Pull request by Geometrically**
Switches to trolley as our payout provider.

Fixes MOD-530

### CODE REVIEW

**Review by OmegaJak — COMMENTED** [review]
Looks good, just minor nitpicks about names/organization largely.

File: src/queue/payouts.rs [hunk 1]
```diff
@@ -1,203 +1,236 @@
...
-                href: String,
-            }
+        Ok((res.batch.id, payment_id))
+    }
 
-            #[derive(Deserialize)]
-            struct PayoutsResponse {
-                pub links: Vec<PayPalLink>,
-            }
+    // listen to webhooks for batch status, account status
```
[the remark this answers is not in the document]
Comment (Geometrically) [reply src/queue/payouts.rs:184]: this was just a todo comment for myself when I was implementing this

File: src/models/projects.rs [hunk 2]
```diff
@@ -247,7 +247,7 @@ impl SideType {
         }
     }
 
-    pub fn from_str(string: &str) -> SideType {
+    pub fn from_string(string: &str) -> SideType {
```
[the remark this answers is not in the document]
Comment (Geometrically) [reply src/models/projects.rs:250]: Clippy lint error, says `from_str` will be confused with FromStr trait impl

File: src/queue/payouts.rs [hunk 3]
```diff
@@ -1,203 +1,236 @@
...
+                }
 
-        Ok(())
+                return Err(ApiError::Payments(
+                    "could not retrieve Trolley error body".to_string(),
+                ));
+            }
+        }
+
+        Ok(serde_json::from_value(value)?)
```
[the remark this answers is not in the document]
Comment (Geometrically) [reply src/queue/payouts.rs:124]: It wouldn't be called anywhere else, and where it is in describes what it's doing without any more complexity. Don't really see a point of having 2 separate methods tbh

File: src/models/users.rs [hunk 4]
```diff
@@ -201,3 +134,89 @@ impl Role {
...
+    }
+
+    pub fn is_failed(&self) -> bool {
+        match self {
+            PayoutStatus::Pending => false,
+            PayoutStatus::Failed => true,
+            PayoutStatus::Processed => false,
+            PayoutStatus::Returned => true,
+            PayoutStatus::Processing => false,
+        }
```
[the remark this answers is not in the document]
Comment (Geometrically) [reply src/models/users.rs:220]: personal pref

File: src/queue/payouts.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (OmegaJak) [reply src/queue/payouts.rs:124]: Most of what we do with our code is read it, so it is very important that the code be easy to read. Pulling giant self-contained chunks of work into methods allows us to more easily read and understand what the top level method is doing. If we care, we can then look at what the lower method is doing. If the methods are sensibly separated, at each level you mostly only have to keep that method's code in your head. This all makes the code easier to read and thus easier to maintain.

Methods also help us get away from very deeply indented code, which gives us less and less screen real estate as things get more nested.

Methods also give us opportunities to introduce meaningful names that provide further context to the code, describing what it does (without the need for comments, which have a tendency to lie over time).

They also give us opportunities to poke the code in lower-level tests than full-scale integration tests.

**Review by OmegaJak — APPROVED** [review]

File: src/queue/payouts.rs [hunk 5]
```diff
@@ -1,203 +1,332 @@
...
-        payout.amount.value -= fee;
-        payout.amount.value = payout.amount.value.round_dp(2);
+        let fee = self.get_estimated_fees(recipient, amount).await?;
 
-        if payout.amount.value <= Decimal::ZERO {
-            return Err(ApiError::InvalidInput(
-                "You do not have enough funds to make this payout!".to_string(),
+        if fee.estimated_fees > amount || fee.route_minimum > amount {
+            return Err(ApiError::Payments(
+                "Account balance is to low to withdraw funds".to_string(),
```
Comment (OmegaJak) [inline src/queue/payouts.rs:166]:
```suggestion
                "Account balance is too low to withdraw funds".to_string(),
```

**Pull request merged**